### ISSUE TITLE
Processes: Integration

### ISSUE THREAD

**Issue by bryanjos** [bryanjos/elixirscript#99]
I made a separate processes repo for the js processes library. It does a pretty good job of mimicking erlang processes. This will be included in the elixirscript repo. 
https://github.com/bryanjos/processes

This also has an implementation of a GenServer as well.

It works well in JavaScript, but need to figure out a way of making it work for elixirscript. The biggest issue being how should functions be translated now? Also, what does it mean for interoperability with JavaScript (#170)? For example, we could turn all functions into generator functions, but how would that work when passing in a function from elixirscript as a callback to a JavaScript function?

**Comment by bryanjos** [comment]
Started working on this in the [generators branch](https://github.com/bryanjos/elixirscript/tree/generators)

**Comment by bryanjos** [comment]
This may go on the back burner for a while. Going through and changing everything to generators, things just didn't feel right. Operability with JavaScript libraries kind of goes out the window because a lot of them expect functions and not generators for callbacks. I don't think that is worth it.

So maybe back to the drawing board on this? Or maybe JavaScript interoperability can be reassessed?

**Comment by crertel** [comment]
In my mind, problem would be...what happens if an exception is thrown in JS land? What was your approach to handling that.

**Comment by bryanjos** [comment]
In the implementation above, the generator that is the process is wrapped in a try catch. If an exception is thrown, it's caught and goes through the flow of terminating the process.

**Comment by bryanjos** [comment]
An alternative is webworkers. They have most of the same properties of Erlang processes. The issues that would leave are how to handle importing scripts, as in how to handle which scripts should be imported into the web worker. Web workers can access the Dom so there would need to be special care for that as well. And also transferring any context if needed.

The import scripts part is complicated by the fact that js has so many module systems. Elixirscript targets es6 modules, but these still need to be transpiled, which may effect what and how scripts are imported. For instance, if using system.js, you would need to import system.js. and then there is the pathing problem.

Tldr; I really wish web workers could be the solution.

**Comment by crertel** [comment]
So, my first (lazy) attempt at solving that problem would be to have the runtime boot up some number of webworkers with a bootstrap script...just enough to get running and able to eval() a script sent to them via `postMessage`.

That done, you have a webworker pool that is sending messages around. The main worker thread (GUI) is where the DOM is manipulated. Comedy option: route messages that signify that they'll change the DOM to the GUI thread, where they are applied (via React or whatever).

The performance is a bit gross, but it should function as a proof-of-concept.

**Comment by bryanjos** [comment]
Your comedy option is a decent one. I've thought the same actually. Using a virtual-dom to manipulate in the web workers and then sending messages to the special UI thread to update the UI. It's not all that different from the design of a lot of mobile and native desktop UI frameworks.

I'm just not sure about the pool of web workers. I think I need an example. I've also been thinking if this is the way to go, then maybe forcing the compilation to a certain module system to make sure it works as expected.

**Comment by ijcd** [comment]
So what does clojurescript do in core.async and for js interop? It's similar with immutability and all. Seems like it could be used for inspiration.

**Comment by bryanjos** [comment]
It looks like clojurescript expects JavaScript code to be at the global namespace for interop. That wouldn't really work here. Looking at Erlang ports and/or Elm ports for inspiration is probably the best way to think about JavaScript interop when processes are implemented.

**Comment by crertel** [comment]
Well, for interop, one could namespace the normal JS functions under a process (perhaps even with catch/throw internally to protect against exceptions?).

Absolutely worst case, perhaps a process with a `JSI` (JavaScriptInterop) that takes a string or strings to dereference and find an object in the global namespaces.

Something like:

`JSI.cast('window','alert','wtf dude');`

?

**Comment by bryanjos** [comment]
What you first describe is pretty close to what I think part of the solution would be. If there was an exception in the JavaScript port, it would be contained to the process calling the port.

Here are some links to ports in Erlang, Elixir, and Elm:
http://erlang.org/doc/reference_manual/ports.html
http://theerlangelist.com/article/outside_elixir
http://elm-lang.org/guide/interop

**Comment by RangerMauve** [comment]
In terms of generator functions, what if you model individual processes as generators or [async-functions](https://github.com/tc39/ecmascript-asyncawait)?

**Comment by bryanjos** [comment]
Actually that's 100% how the implementation above works :). It's just to have the level of control needed to mimic the Erlang scheduler, all the functions would need to be generators.

I'm open to any alternatives. A new set of eyes on this would be awesome. I may be looking at it in tunnel vision

**Comment by RangerMauve** [comment]
So, is there a hard dependency on the library then you linked to earlier? It seems like using async functions for anything that wants to do send/receive calls, or any IO would play pretty nice with JS. Would there maybe be a way to detect which functions don't make use of those features so that the compiler will automagically declare functions that don't need to be async the way they normally do, and then just make heavy use of `await` everywhere in the generated code? Obviously if you went with async functions, you'd need to have a transpiler step with something like Babel in order to make it run in most browsers. Async functions are easier to incorporate with regular functions because they just act like functions that return promises from the outside. And inside you can use `await` to suspend execution until a promise resolves, or if you await on a non-promise (like the result of a regular function), it'll basically do nothing.

**Comment by bryanjos** [comment]
There isn't a hard dependency on the library in the description. The only hard requirement would be that they act and function like processes in the beam VM. If there are some sacrifices to the properties of a process to get things working overall that's an option as well.

**Comment by RangerMauve** [comment]
Interesting. I'm still pretty new to Elixir, but I've got a lot of modern JS under my belt, so I might be able to help some more once I have my head wrapped over all this. So far it doesn't sound __too__ hard. Are there any docs on how the JS side of things is getting generated?

**Comment by bryanjos** [comment]
Under the hood it uses the [estree](https://github.com/bryanjos/elixir-estree) library to build up the JavaScript AST and generate JavaScript code. In elixirscript, the translations from Elixir AST to JavaScript AST happens in [here](https://github.com/bryanjos/elixirscript/blob/master/lib/elixir_script/translator.ex) and [here](https://github.com/bryanjos/elixirscript/tree/master/lib/elixir_script/translator)

There is also some JavaScript code that is used and copied along with the generated JavaScript [here](https://github.com/bryanjos/elixirscript/tree/master/src/javascript)

And some reading that may help understand how processes work:
http://erlang.org/doc/reference_manual/processes.html
https://joearms.github.io/2013/04/02/Red-and-Green-Callbacks.html
http://elixir-lang.org/docs/stable/elixir/Kernel.SpecialForms.html#receive/1

**Comment by RangerMauve** [comment]
Do you happen to know which operations will be blocking in advance in this case?

**Comment by bryanjos** [comment]
So nothing blocks except `receive` which blocks the process it's on either infinitely or for a set amount of time. 
If receive receives a message that matches a pattern set on it, the system blocks until that code is finished.

**Comment by RangerMauve** [comment]
Perfect. I assume the other stuff that would need attention is APIs wrapping over top of JS non-blocking APIs, like doing HTTP requests. Though I'm not even sure how that would work unless there was a custom library just for ElixirScript.

Can receive be called within anonymous functions or blocks of code? Or would it just be in the top level scope for a process? If it's the latter than that means you'd only need to convert functions that will be used as processes as async functions and leave everything else as vanilla JS functions.

**Comment by bryanjos** [comment]
There is another issue for defining JavaScript interoperability. `receive` can be called within anonymous functions and blocks of code.

**Comment by RangerMauve** [comment]
So, is it safe to say that anonymous functions contain `receive` in a closure when they're passed around? Maybe you'd need to assume that everything is an async function for the STD, then. :(

Maybe it could look something like:

```javascript
spawn(async function(self, send, receive){
  var sum = await receive(async function(message){
     if(message[0]==="add"){
       var a = message[1];
       var b = message[2];
       return a + b;
     } if(message[0]==="about_to_send_something"){
       return await receive(function(){ /* Doesn't need to be async because it doesn't use receive */});
     } else {
       // Whatever error
     }
  });

  send(somePID, ["result", sum]);
});
```

Probably a convoluted example where recieve gets used within a block of code passed to a standard library method:

```javascript
Enum.map(async function(item){
  return await receive(function(message){
    if(message[0]==="something"){
      var data = message[1];
      return Map.merge(item, {data: data});
    } else {
      // Whatever error
    }
  })
});
```

Obviously this is pretty gross, but that's sort of the dream code I'm thinking about

**Comment by bryanjos** [comment]
async/await may not work well because there isn't enough control over the flow. Generators work because they are lower level and you can control when a generator continues.

I'm going to try to work on this over the next couple of months. Hopefully by the end of the summer processes are working in one shape or form.

**Comment by bryanjos** [comment]
Implementing Port probably can come before or be implemented at the same time as Processes

**Comment by bryanjos** [comment]
This is/will be resolved with #184
